### ISSUE TITLE
Syntax error when discovering tests

### ISSUE THREAD

**Issue by matprzybyla** [hbenl/vscode-mocha-test-adapter#68]
Hi,

We are using some third party lib that outputs using newer module syntax. 
To compile and run tests from CMD we added `esm` module to out mocha --require and this helped mocha to crunch through module definitions. Othewise it could not understand some exports.

Now the same issue is with MochaTextExplorer - it shows the same behavior when discovering tests:
```
export * from './Async';
^^^^^^
SyntaxError: Unexpected token export
```

Is there a way to add `esm` module to pipeline somehow? Or make exporer understand such syntax? We tried using "mochaExplorer.require": "esm" but it fails to find this module.

**Comment by hbenl** [comment]
> We tried using "mochaExplorer.require": "esm" but it fails to find this module.

That's exactly what you need to do to use `esm` in your tests. Where is the `esm` package installed on your machine? Can you give me your full mocha config for the command line (i.e. the exact command that you use to invoke mocha and mocha config files, if you have any) and for VS Code? Furthermore, you could use the `mochaExplorer.logfile` or `mochaExplorer.logpanel` settings to create a log and attach it here.

**Comment by matprzybyla** [comment]
I think I was able to move a bit further, but stuck on another problem. 

We are using workspace in VSCode that has several projects: A, B, C. Those are made in different times using newer or older approach to some things. Still all using mocha.
The workspace file is locatetd in root above project folders.

Sample extract from package.json from one project looks like this
`"test": "mocha -r jsdom-global/register --require esm --require ignore-styles --file testsSetup.js clientAppDist/tests/**/*.js",`

Workspace config looks like this
```
		"mochaExplorer.logpanel": true,
		"mochaExplorer.files": "clientAppDist/tests/**/*.js",
		"mochaExplorer.mochaPath": "node_modules/mocha",
		"mochaExplorer.require": ["jsdom-global/register", "esm", "source-map-support/register"],
```

I'm geting 2 error messages in mocha test explorer - I guest for 2 projects. 
One is `TypeError: window.requestAnimationFrame is not a function` - this requires polyfill on test start that is in `testsSetup.js` when run from cmd line (3rd party lib requires it).
One error is missing jsdom `Error: Cannot find module 'jsdom-global/register'` as one of the projects does not use it.

Questions are: 
- can I somehow configure test explorer per project? 
- can I add "tests startup" file to test explorer that adds required polyfills

**Comment by hbenl** [comment]
> can I somehow configure test explorer per project?

Yes, if you use a [multi-root workspace](https://code.visualstudio.com/docs/editor/multi-root-workspaces) you can add VS Code settings for each root folder.

> can I add "tests startup" file to test explorer that adds required polyfills

Yes, but this is currently only possible using a mocha configuration file. I'd recommend creating a `test/mocha.opts` file containing your mocha command line args:
```
-r jsdom-global/register
--require esm
--require ignore-styles
--file testsSetup.js
clientAppDist/tests/**/*.js
```
You should then be able to call `mocha` without arguments and it should pick up the arguments from the opts file. Mocha Explorer will also read that file so you can remove the `mochaExplorer.files` and `mochaExplorer.require` settings. It will also use the `--file testsSetup.js` setting.

**Comment by matprzybyla** [comment]
@hbenl Thanks for support.
I was able to run all needed tests in all projects using separate mocha.opts files. :)

Still having one problem left:
In one of projects I cannot see the source code (tests are running). Using source-map-support and generating source maps but no luck. The only "unique" thing with this project is that we have separate tsconfig.json file to build tests - may this have some influence? Does MochaTestExporer touch tsconfig.json somehow or it is transaprent?

**Comment by hbenl** [comment]
> Does MochaTestExporer touch tsconfig.json somehow or it is transaprent?

It's completely transparent - the Mocha Explorer doesn't know anything about Typescript (and hence `tsconfig.json`), it just uses the sourcemaps if you add `--require source-map-support/register` to `mocha.opts`.
Another possibility (which I recommend) is to use `ts-node` instead: add `--require ts-node/register` to your `mocha.opts` and change `tests/**/*.js` to `tests/**/*.ts`. Then Mocha Explorer will use the Typescript files directly (no need to watch and transpile them anymore). But in this case you'd probably have to figure out how to make `ts-node` use the `tsconfig.json` for the tests (instead of the default `tsconfig.json`).
To figure out why it isn't working you could look for messages like `Worker: Looking for [...] in Error:` followed by a stack trace in the log: that's how Mocha Explorer finds the position of your test in your source file.
